File: workers/PaymentWorker.js
import pool from "../db.js";
import { isAlreadyDone, markDone } from "../core/workerIdempotency.js";

function calculateDueDate(days) {
  const date = new Date();
  date.setDate(date.getDate() + Number(days));
  return date;
}

export async function execute(context) {

  const { invoice_id, organization_id, config } = context;

  if (!invoice_id || !organization_id) {
    return { success: false, reason: "Missing invoice context" };
  }

  // Clear idempotency so PaymentWorker always reruns fresh.
  // Invoice can reach PAYMENT_READY multiple times (normal path + exception resolve path).
  // Each time it must reschedule payment and reset the approval workflow.
  await pool.query(
    `DELETE FROM worker_completion_log
     WHERE invoice_id      = $1
       AND organization_id = $2
       AND state           = 'PAYMENT_READY'`,
    [invoice_id, organization_id]
  );

  const stateRes = await pool.query(
    `SELECT current_state
     FROM invoice_state_machine
     WHERE invoice_id      = $1
       AND organization_id = $2`,
    [invoice_id, organization_id]
  );

  if (!stateRes.rows.length) {
    return { success: false, reason: "State record not found" };
  }

  if (stateRes.rows[0].current_state !== "PAYMENT_READY") {
    return { success: false, reason: "Invalid state for payment scheduling" };
  }

  const invoiceRes = await pool.query(
    `SELECT data
     FROM invoice_extracted_data
     WHERE invoice_id      = $1
       AND organization_id = $2`,
    [invoice_id, organization_id]
  );

  if (!invoiceRes.rows.length) {
    return { success: false, reason: "Invoice data not found" };
  }

  const invoice       = invoiceRes.rows[0].data || {};
  const paymentPolicy = config?.payment || {};

  let dueDate;

  if (invoice?.due_date) {
    dueDate = new Date(invoice.due_date);
  } else if (paymentPolicy?.default_due_days) {
    dueDate = calculateDueDate(paymentPolicy.default_due_days);
  } else {
    return { success: false, reason: "Payment due date missing" };
  }

  const paymentMethod =
    invoice?.payment_method ||
    paymentPolicy?.default_payment_method ||
    "BANK_TRANSFER";

  await pool.query(
    `INSERT INTO invoice_payment_schedule
      (invoice_id, organization_id,
       payment_status, payment_due_date,
       payment_method, scheduled_at)
    VALUES ($1,$2,$3,$4,$5,NOW())
    ON CONFLICT (invoice_id, organization_id)
    DO UPDATE SET
      payment_status   = EXCLUDED.payment_status,
      payment_due_date = EXCLUDED.payment_due_date,
      payment_method   = EXCLUDED.payment_method`,
    [invoice_id, organization_id, "SCHEDULED", dueDate, paymentMethod]
  );

  if (!config?.approval?.levels?.length) {
    return { success: false, reason: "Approval levels not configured" };
  }

  const invoiceTotal   = parseFloat(invoice.total_amount || 0);
  const levels         = [...config.approval.levels].sort((a, b) => b.min_amount - a.min_amount);
  const matched        = levels.find(l => invoiceTotal >= l.min_amount);
  const required_level = matched?.approver_role || "CFO";

  // Reset approval workflow for this payment cycle
  await pool.query(
    `INSERT INTO invoice_approval_workflow
       (invoice_id, organization_id, approval_level,
        required_approval_level, approval_status, created_at)
     VALUES ($1, $2, $3, $4, 'PENDING', NOW())
     ON CONFLICT (invoice_id, organization_id)
     DO UPDATE SET
       approval_level          = EXCLUDED.approval_level,
       required_approval_level = EXCLUDED.required_approval_level,
       approval_status         = 'PENDING',
       decision_at             = NULL`,
    [invoice_id, organization_id, required_level, required_level]
  );

  await markDone(invoice_id, organization_id, "PAYMENT_READY");

  return { success: true, nextState: "PENDING_APPROVAL" };
}

File: python/workers/payment_worker.py
from db import get_connection
from datetime import datetime, timedelta

def run(state: dict) -> dict:
    invoice_id = state["invoice_id"]
    organization_id = state["organization_id"]
    config = state.get("config", {})

    # check state is PAYMENT_READY
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        "SELECT current_state FROM invoice_state_machine WHERE invoice_id = %s AND organization_id = %s",
        (invoice_id, organization_id)
    )
    row = cur.fetchone()
    cur.close()
    conn.close()

    if not row or row[0] != "PAYMENT_READY":
        return {**state, "next_state": None, "reason": "Wrong state for payment scheduling"}

    # get extracted invoice data
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        "SELECT data FROM invoice_extracted_data WHERE invoice_id = %s AND organization_id = %s",
        (invoice_id, organization_id)
    )
    row = cur.fetchone()
    cur.close()
    conn.close()

    if not row:
        return {**state, "next_state": None, "reason": "Extracted data not found"}

    invoice = row[0]
    payment_config = config.get("payment", {})

    # calculate due date
    due_date = None

    if invoice.get("due_date"):
        try:
            due_date = datetime.strptime(str(invoice["due_date"]), "%Y-%m-%d")
        except:
            pass

    if not due_date:
        default_days = payment_config.get("default_due_days", 30)
        due_date = datetime.now() + timedelta(days=default_days)

    payment_method = (
        invoice.get("payment_method") or
        payment_config.get("default_payment_method") or
        "BANK_TRANSFER"
    )

    # get approval tier for this invoice amount
    total_amount = float(invoice.get("total_amount") or 0)
    levels = config.get("approval", {}).get("levels", [])
    levels_sorted = sorted(levels, key=lambda x: x["min_amount"], reverse=True)
    matched_level = next(
        (l for l in levels_sorted if total_amount >= l["min_amount"]),
        None
    )

    if not matched_level:
        return {**state, "next_state": None, "reason": "No matching approval tier found"}

    required_approval_level = matched_level["approver_role"]

    # save payment schedule
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        """
        INSERT INTO invoice_payment_schedule
            (invoice_id, organization_id, payment_status,
             payment_due_date, payment_method, scheduled_at)
        VALUES (%s, %s, %s, %s, %s, NOW())
        ON CONFLICT (invoice_id, organization_id)
        DO UPDATE SET
            payment_status = EXCLUDED.payment_status,
            payment_due_date = EXCLUDED.payment_due_date,
            payment_method = EXCLUDED.payment_method
        """,
        (invoice_id, organization_id, "SCHEDULED", due_date, payment_method)
    )
    conn.commit()
    cur.close()
    conn.close()

    # save approval workflow
    conn = get_connection()
    cur = conn.cursor()
    cur.execute(
        """
        INSERT INTO invoice_approval_workflow
            (invoice_id, organization_id, approval_level,
             required_approval_level, approval_status, created_at)
        VALUES (%s, %s, %s, %s, 'PENDING', NOW())
        ON CONFLICT (invoice_id, organization_id)
        DO UPDATE SET
            approval_level = EXCLUDED.approval_level,
            required_approval_level = EXCLUDED.required_approval_level,
            approval_status = EXCLUDED.approval_status
        """,
        (invoice_id, organization_id, required_approval_level, required_approval_level)
    )
    conn.commit()
    cur.close()
    conn.close()

    return {**state, "next_state": "PENDING_APPROVAL", "reason": f"Payment scheduled. Requires approval from: {required_approval_level}"}